_stbt/core.py

```diff
@@ -1948,28 +1946,25 @@ def pull_frame(self, timeout_secs=10):
 
     def frames(self, timeout_secs=None):
         import time
-        self.start_timestamp = None
+        if timeout_secs is not None:
+            end_time = time.time() + timeout_secs
 
         with self.lock:
             while True:
                 ddebug("user thread: Getting sample at %s" % time.time())
-                sample = self._get_sample(max(10, timeout_secs))
+                sample = self._get_sample(max(10, timeout_secs or 0))
                 ddebug("user thread: Got sample at %s" % time.time())
                 timestamp = sample.time
 
-                if timeout_secs is not None:
-                    if not self.start_timestamp:
-                        self.start_timestamp = timestamp
-                    if timestamp - self.start_timestamp > timeout_secs:
-                        debug("timed out: %.3f - %.3f > %.3f" % (
-                            timestamp, self.start_timestamp, timeout_secs))
-                        return
-
                 try:
                     yield array_from_sample(sample)
                 finally:
                     self._sink_pipeline.push_sample(sample)
 
+                if timeout_secs is not None and timestamp > end_time:
+                    debug("timed out: %.3f > %.3f" % (timestamp, end_time))
+                    return
+
```

Comment:
So now we're checking for timeout *after* yielding the frame -- also a change in behaviour. It's kind of reverted by 1bde345 on branch `one-past-end`.

Reply:
I've fixed this in a fixup.